src/tiff.ml

```diff
@@ -920,25 +928,37 @@ module Data = struct
 
       (* We iterate through every strip *)
       for strip = first_strip to last_strip - 1 do
-        let raw_strip_buffer = Cstruct.create strip_bytecounts.(strip) in
-        let strip_offset = Optint.Int63.of_int strip_offsets.(strip) in
-
-        (* Fill the strip buffer *)
-        ro ~file_offset:strip_offset [ raw_strip_buffer ];
-
-        let expected_size = rows_per_strip * width * bytes_per_pixel in
-
         let strip_buffer =
-          match Ifd.compression ifd with
-          | No_compression ->
-              if Cstruct.length raw_strip_buffer < expected_size then
-                failwith "Strip is unexpectedly short";
-              raw_strip_buffer
-          | LZW ->
-              let uncompressed_buffer = Cstruct.create expected_size in
-              Lzw.decode raw_strip_buffer uncompressed_buffer;
-              uncompressed_buffer
-          | _ -> failwith "Unsupported compression"
+          match List.assoc_opt strip t.strip_cache with
+          | Some strip -> strip
+          | None ->
+              let raw_strip_buffer = Cstruct.create strip_bytecounts.(strip) in
+              let strip_offset = Optint.Int63.of_int strip_offsets.(strip) in
+
+              (* Fill the strip buffer *)
+              ro ~file_offset:strip_offset [ raw_strip_buffer ];
+
+              let expected_size = rows_per_strip * width * bytes_per_pixel in
+
+              let strip_buffer =
+                match Ifd.compression ifd with
+                | No_compression ->
+                    if Cstruct.length raw_strip_buffer < expected_size then
+                      failwith "Strip is unexpectedly short";
+                    raw_strip_buffer
+                | LZW ->
+                    let uncompressed_buffer = Cstruct.create expected_size in
+                    Lzw.decode raw_strip_buffer uncompressed_buffer;
+                    uncompressed_buffer
+                | _ -> failwith "Unsupported compression"
+              in
+
+              t.strip_cache <-
```

Comment:
I'm not exactly sure what this gets compiled to, but I think it would be better to put the matching statement outside of the assignment operation. IDK if the compiler is smart enough to recognise that if t.caching_policy = `NoCaching`is a no-op.